packages/fxa-settings/src/pages/PostVerify/ThirdPartyAuthCallback/index.tsx

```diff
@@ -25,143 +36,134 @@ type LinkedAccountData = {
-    if (redirectURL) {
-      // get the stashed state with origin information
-      // use it to reconstruct redirect for oauth
-      // the state is the entire URL of the origin and includes the redirect_uri in a param
-      // if authenticating from a RP
-      const state = params.get('state');
-      if (state) {
-        // we may need to deconstruct the state to access/modify the redirect URL
-        const stateParams = new URL(decodeURIComponent(state)).searchParams;
-        const redirect = stateParams.get('redirect_uri');
-        // if the state contains a redirect_uri, we need to redirect to RP
-        // otherwise we redirect internally
-        if (redirect) {
-          hardNavigate(redirect);
-        } else {
-          // general redirect to settings for non-RP
-          // currently, redirect to /settings fails with an "unauthenticated" error from GQL
-          // and redirects to /signin (on backbone) where ThirdPArty Auth successfully
-          // navigates to /settings
-          navigate('/settings');
-        }
-      }
+  const integration = useIntegration();
+  const authClient = useAuthClient();
+  const location = useLocation();
+
+  const linkedAccountData = useRef({} as LinkedAccountData);
+
+  const { finishOAuthFlowHandler } = useFinishOAuthFlowHandler(
+    authClient,
+    integration || ({} as Integration)
+  );
+
+  const storeLinkedAccountData = useCallback(
+    async (linkedAccount: LinkedAccountData) => {
+      const accountData: StoredAccountData = {
+        email: linkedAccount.email,
+        uid: linkedAccount.uid,
+        lastLogin: Date.now(),
+        sessionToken: linkedAccount.sessionToken,
+        verified: true,
+        metricsEnabled: true,
+      };
+      return storeAccountData(accountData);
+    },
+    []
+  );
+
+  const verifyThirdPartyAuthResponse = useCallback(async () => {
+    const { code: thirdPartyOAuthCode, provider } = (
+      integration as ThirdPartyAuthCallbackIntegration
+    ).thirdPartyAuthParams();
+
+    if (!thirdPartyOAuthCode) {
+      return hardNavigate('/');
     }
-  };
-
-  // auth params are received from the third party auth provider
-  // and are required to verify the account
-  const getAuthParams = () => {
-    const code = params.get('code');
-    const providerFromParams = params.get('provider');
-    let provider: AUTH_PROVIDER | undefined;
-    if (providerFromParams === 'apple') {
-      provider = AUTH_PROVIDER.APPLE;
-    } else {
-      provider = AUTH_PROVIDER.GOOGLE;
+
+    try {
+      const linkedAccount: LinkedAccountData =
+        await account.verifyAccountThirdParty(
+          thirdPartyOAuthCode,
+          provider,
+          undefined,
+          queryParamsToMetricsContext(
+            flowQueryParams as unknown as Record<string, string>
+          )
+        );
+      await storeLinkedAccountData(linkedAccount);
+
+      linkedAccountData.current = linkedAccount;
+
+      const fxaParams = (
```

Comment:
It'd be better to use the typeguard here than casting which isn't really type safe. `isThirdPartyAuthCallbackIntegration`